src/bio/bio_device.c

```diff
@@ -689,114 +805,59 @@ bio_set_led_state(struct bio_xs_context *xs_ctxt, uuid_t dev_uuid,
-		D_GOTO(skip_led_str, rc = 0);
+		g_opts.action = LED_ACTION_RESET;
+		g_opts.led_state = bio_dev->bb_led_state;
+		goto skip_led_str;
 	}
 
-	if (led_state == NULL)
+	if (led_state_str == NULL)
 		return -DER_INVAL;
 
 	/* Determine SPDK LED state based on led_state string */
-	if (strcasecmp(led_state, "identify") == 0) {
-		new_led_state = SPDK_VMD_LED_STATE_IDENTIFY;
-	} else if (strcasecmp(led_state, "on") == 0) {
-		new_led_state = SPDK_VMD_LED_STATE_FAULT;
-	} else if (strcasecmp(led_state, "fault") == 0) {
-		new_led_state = SPDK_VMD_LED_STATE_FAULT;
-	} else if (strcasecmp(led_state, "off") == 0) {
-		new_led_state = SPDK_VMD_LED_STATE_OFF;
-	} else {
+	for (led_state = SPDK_VMD_LED_STATE_OFF;
+	     led_state <= SPDK_VMD_LED_STATE_REBUILD;
+	     led_state++) {
+		if (strcmp(led_state_str, g_led_states[led_state]) == 0) {
+			g_opts.led_state = (enum spdk_vmd_led_state)led_state;
+			break;
+		}
+	}
+
+skip_led_str:
+	if (g_opts.led_state == SPDK_VMD_LED_STATE_UNKNOWN) {
 		D_ERROR("LED state is not valid or supported\n");
 		return -DER_NOSYS;
 	}
 
-skip_led_str:
 	rc = fill_in_traddr(&b_info, bio_dev->bb_name);
 	if (rc) {
 		D_ERROR("Unable to get traddr for device:%s\n",
 			bio_dev->bb_name);
 		return -DER_INVAL;
 	}
 
-
-	if (spdk_pci_addr_parse(&pci_addr, b_info.bdi_traddr)) {
+	if (spdk_pci_addr_parse(&g_opts.pci_addr, b_info.bdi_traddr)) {
 		D_ERROR("Unable to parse PCI address: %s\n", b_info.bdi_traddr);
 		D_GOTO(free_traddr, rc = -DER_INVAL);
 	}
 
-	for (pci_device = spdk_pci_get_first_device(); pci_device != NULL;
-	     pci_device = spdk_pci_get_next_device(pci_device)) {
-		if (spdk_pci_addr_compare(&pci_addr, &pci_device->addr) == 0) {
-			found = true;
-			break;
-		}
-	}
-
-	if (found) {
-		if (strcmp(spdk_pci_device_get_type(pci_device), "vmd") != 0) {
-			D_DEBUG(DB_MGMT, "%s is not a VMD device\n",
-				b_info.bdi_traddr);
-			D_GOTO(free_traddr, rc = -DER_NOSYS);
-		}
-	} else {
-		D_ERROR("Unable to set led state, VMD device not found\n");
-		D_GOTO(free_traddr, rc = -DER_INVAL);
-	}
-
-	/* First check the current state of the VMD LED */
-	rc = spdk_vmd_get_led_state(pci_device, &current_led_state);
-	if (rc) {
-		D_ERROR("Failed to get the VMD LED state\n");
-		D_GOTO(free_traddr, rc = -DER_INVAL);
-	}
-
-	/* If the current state of a device is FAULTY we do not want to reset */
-	if ((current_led_state == SPDK_VMD_LED_STATE_FAULT) && reset)
-		D_GOTO(state_set, rc);
+	g_status = 0;
 
-	if (!reset)
-		D_DEBUG(DB_MGMT, "Setting VMD device:%s LED state to %s(%d)\n",
-			b_info.bdi_traddr, led_state, new_led_state);
-	else
-		D_DEBUG(DB_MGMT, "Resetting VMD device:%s LED state to %d\n",
-			b_info.bdi_traddr, bio_dev->bb_led_state);
-
-	/* Save the current state in bio_bdev, will be restored by init xs */
-	if (!reset)
-		bio_dev->bb_led_state = current_led_state;
-
-	if (current_led_state == new_led_state)
-		D_GOTO(state_set, rc);
+	spdk_pci_for_each_device(NULL, led_device_action);
```

Comment:
Minor: all devices will be iterated no matter if the requested device has been found and processed?